File: python/two_sum.py
from typing import List


# Problem link: https://leetcode.com/problems/two-sum/
# Complexity: Time of O(n) and memory O(n).
def two_sum(numbers: List[int], target: int) -> List[int]:
    previous_numbers = {}  # number : index

    for index, number in enumerate(numbers):
        difference = target - number

        if difference in previous_numbers:
            return [previous_numbers[difference], index]

        previous_numbers[number] = index

    # As we are ensured that there is a solution, we expect that the code does not reach here.


class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        return two_sum(nums, target)

File: java/src/main/TwoSum.java
package main;

import java.util.HashMap;

class Solution {

  public int[] twoSum(int[] nums, int target) {
    var previousNumbers = new HashMap<Integer, Integer>();
    int[] output = new int[2];

    for (int i = 0; i < nums.length; i++) {
      int difference = target - nums[i];
      Integer differenceIndex = previousNumbers.get(difference);

      if (differenceIndex != null) {
        output[0] = i;
        output[1] = differenceIndex;
      }

      previousNumbers.put(nums[i], i);
    }

    return output;
  }

  public static void main(String[] args){
    new Solution().twoSum(new int[]{2,7,11,15}, 9);
  }
}
